mmengine/utils/dl_utils/setup_env.py

```diff
@@ -39,23 +46,26 @@ def set_multi_processing(mp_start_method: str = 'fork',
     except ImportError:
         pass
 
-    # setup OMP threads
-    # This code is referred from https://github.com/pytorch/pytorch/blob/master/torch/distributed/run.py  # noqa
-    if 'OMP_NUM_THREADS' not in os.environ and distributed:
-        omp_num_threads = 1
-        warnings.warn(
-            'Setting OMP_NUM_THREADS environment variable for each process'
-            f' to be {omp_num_threads} in default, to avoid your system '
-            'being overloaded, please further tune the variable for '
-            'optimal performance in your application as needed.')
-        os.environ['OMP_NUM_THREADS'] = str(omp_num_threads)
-
-    # setup MKL threads
-    if 'MKL_NUM_THREADS' not in os.environ and distributed:
-        mkl_num_threads = 1
-        warnings.warn(
-            'Setting MKL_NUM_THREADS environment variable for each process'
-            f' to be {mkl_num_threads} in default, to avoid your system '
-            'being overloaded, please further tune the variable for '
-            'optimal performance in your application as needed.')
-        os.environ['MKL_NUM_THREADS'] = str(mkl_num_threads)
+    if distributed:
+        # setup OMP threads
+        # This code is referred from https://github.com/pytorch/pytorch/blob/master/torch/distributed/run.py  # noqa
+        if omp_num_threads is None and 'OMP_NUM_THREADS' not in os.environ:
+            omp_num_threads = 1
+            warnings.warn(
+                'Setting OMP_NUM_THREADS environment variable for each process'
+                f' to be {omp_num_threads} in default, to avoid your system '
+                'being overloaded, please further tune the variable for '
+                'optimal performance in your application as needed.')
+        if omp_num_threads is not None:
+            os.environ['OMP_NUM_THREADS'] = str(omp_num_threads)
```

Comment:
```suggestion
    if not distributed:
        return 

    if omp_num_threads is not None:
        os.environ['OMP_NUM_THREADS'] = str(omp_num_threads)
    else:
        omp_num_threads = 1
        warnings.warn(
            'Setting OMP_NUM_THREADS environment variable for each process'
            f' to be {omp_num_threads} in default, to avoid your system '
            'being overloaded, please further tune the variable for '
            'optimal performance in your application as needed.')
    
```

Reply:
Should we set OMP_NUM_THREADS and KML_NUM_THREADS when not distributed?
How about change whole thing to:
```python
    # setup OMP threads
    # This code is referred from https://github.com/pytorch/pytorch/blob/master/torch/distributed/run.py  # noqa
    if omp_num_threads is None and 'OMP_NUM_THREADS' not in os.environ and distributed:
        omp_num_threads = 1
        warnings.warn(
            'Setting OMP_NUM_THREADS environment variable for each process'
            f' to be {omp_num_threads} in default, to avoid your system '
            'being overloaded, please further tune the variable for '
            'optimal performance in your application as needed.')
    if omp_num_threads is not None:
        os.environ['OMP_NUM_THREADS'] = str(omp_num_threads)

    # setup MKL threads
    if mkl_num_threads is None and 'MKL_NUM_THREADS' not in os.environ and distributed:
        mkl_num_threads = 1
        warnings.warn(
            'Setting MKL_NUM_THREADS environment variable for each process'
            f' to be {mkl_num_threads} in default, to avoid your system '
            'being overloaded, please further tune the variable for '
            'optimal performance in your application as needed.')
    if mkl_num_threads is not None:
        os.environ['MKL_NUM_THREADS'] = str(mkl_num_threads)
```

Reply:
It seems `omp_num_threads` and  `mkl_num_threads` have the highest priority, how about check `omp_num_threads` is None or not firstly :joy:?  Maybe it could be more readable. In the current implementation, `omp_num_threads` could be checked twice, it seems a little bit weird.

Reply:
Do you mean this?
```python
    if omp_num_threads is None:
        if 'OMP_NUM_THREADS' not in os.environ and distributed:
            os.environ['OMP_NUM_THREADS'] = '1'
            warnings.warn(
                'Setting OMP_NUM_THREADS environment variable for each process'
                ' to be 1 in default, to avoid your system being overloaded, '
                'please further tune the variable for optimal performance in '
                'your application as needed.')
    else:
        os.environ['OMP_NUM_THREADS'] = str(omp_num_threads)

```

Reply:
> Do you mean this?
> 
> ```python
>     if omp_num_threads is None:
>         if 'OMP_NUM_THREADS' not in os.environ and distributed:
>             os.environ['OMP_NUM_THREADS'] = '1'
>             warnings.warn(
>                 'Setting OMP_NUM_THREADS environment variable for each process'
>                 ' to be 1 in default, to avoid your system being overloaded, '
>                 'please further tune the variable for optimal performance in '
>                 'your application as needed.')
>     else:
>         os.environ['OMP_NUM_THREADS'] = str(omp_num_threads)
> ```

LGTM, Is it necessary to set `omp_num_threads` when not distributed?

Reply:
It is hard to say if we do not set `OMP_NUM_THREADS`, the default value will be 1 for 3rd party libraries. 

![image](https://user-images.githubusercontent.com/57566630/196137312-a02f5ec0-5fbb-49bd-8dc0-86e0470dbff2.png)

The main reason we set `OMP_NUM_THREADS=1` here is that if each process enable too many threads during ddp training, system could be overloaded.

Reply:
Well, I'm not very familiar with this (and that's why I leave the code here 😂). I know the origin code here is to avoid overload. But, I add the two param here because I want to get speed up from the multi threads😂. I think we should set OMP_NUM_THREADS even when not distributed, which may bring a considerable speed-up.

Reply:
Alright, I think it is also reasonable. Maybe we can invite @zhouzaida to have a look.